util/buildinfo/buildinfo.go

```diff
@@ -2,136 +2,279 @@ package buildinfo
-		return nil, errors.Wrap(err, "failed to unmarshal buildinfo from config")
+	if format.RemoveAttrs {
+		bi.Attrs = nil
 	}
-	if len(config.BuildInfo) == 0 {
-		return nil, nil
+	if dt, err = json.Marshal(bi); err != nil {
+		return nil, err
 	}
-	var bi []exptypes.BuildInfo
-	if err := json.Unmarshal(config.BuildInfo, &bi); err != nil {
-		return nil, errors.Wrapf(err, "failed to unmarshal %s", ImageConfigField)
+	return dt, nil
+}
+
+var knownAttrs = []string{
+	//"cmdline",
+	"context",
+	"filename",
+	"source",
+
+	//"add-hosts",
+	//"cgroup-parent",
+	//"force-network-mode",
+	//"hostname",
+	//"image-resolve-mode",
+	//"platform",
+	"shm-size",
+	"target",
+	"ulimit",
+}
+
+// filterAttrs filters frontent opt by picking only those that
+// could effectively change the build result.
+func filterAttrs(attrs map[string]*string) map[string]*string {
+	filtered := make(map[string]*string)
+	for k, v := range attrs {
+		if v == nil {
+			continue
+		}
+		// Control args are filtered out
+		if isControlArg(k) {
+			continue
+		}
+		// Always include args and labels
+		if strings.HasPrefix(k, "build-arg:") || strings.HasPrefix(k, "label:") {
+			filtered[k] = v
+			continue
+		}
+		// Filter only for known attributes
+		for _, knownAttr := range knownAttrs {
+			if knownAttr == k {
+				filtered[k] = v
+				break
+			}
+		}
+	}
+	return filtered
+}
+
+var knownControlArgs = []string{
+	"BUILDKIT_CACHE_MOUNT_NS",
+	"BUILDKIT_CONTEXT_KEEP_GIT_DIR",
+	"BUILDKIT_INLINE_BUILDINFO_ATTRS",
+	"BUILDKIT_INLINE_CACHE",
+	"BUILDKIT_MULTI_PLATFORM",
+	"BUILDKIT_SANDBOX_HOSTNAME",
+	"BUILDKIT_SYNTAX",
+}
+
+// isControlArg checks if a build attributes is a control arg
+func isControlArg(attrKey string) bool {
+	for _, k := range knownControlArgs {
+		if strings.HasPrefix(attrKey, "build-arg:"+k) {
+			return true
+		}
+	}
+	return false
+}
+
+// GetMetadata returns buildinfo metadata for the specified key. If the key
+// is already there, result will be merged.
+func GetMetadata(metadata map[string][]byte, key string, reqFrontend string, reqAttrs map[string]string) ([]byte, error) {
+	var (
+		dtbi []byte
+		err  error
+	)
+	if v, ok := metadata[key]; ok {
+		var mbi binfotypes.BuildInfo
+		if errm := json.Unmarshal(v, &mbi); errm != nil {
+			return nil, errors.Wrapf(errm, "failed to unmarshal build info for %q", key)
+		}
+		if reqFrontend != "" {
+			mbi.Frontend = reqFrontend
+		}
+		mbi.Attrs = convertMap(reduceMap(reqAttrs, mbi.Attrs))
+		dtbi, err = json.Marshal(mbi)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to marshal build info for %q", key)
+		}
+	} else {
+		dtbi, err = json.Marshal(binfotypes.BuildInfo{
+			Frontend: reqFrontend,
+			Attrs:    convertMap(reqAttrs),
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to marshal build info for %q", key)
+		}
+	}
+	return dtbi, nil
+}
+
+func reduceMap(original map[string]string, merger map[string]*string) map[string]string {
```

Comment:
The names here flipped I think. The one with the pointers is the original that gets modified. Then you don't need convert as well.